cjdns-hdr/src/data_header.rs

```diff
@@ -0,0 +1,150 @@
+            header_content::ContentType::from(content_number as u32)
+        };
+        Ok(DataHeader { version, content_type })
+    }
+
+    /// Serializes `DataHeader` instance.
+    pub fn serialize(&self) -> Result<Vec<u8>> {
+        if self.version > 15 {
+            return Err(HeaderError::CannotSerialize("invalid header version"));
+        }
+        let mut data_writer = Writer::with_capacity(DATA_HEADER_SIZE);
+        let version_with_flags = if self.version == 0 { HEADER_CURRENT_VERSION << 4 } else { self.version << 4 };
+        let content_type_number = header_content::to_u16(self.content_type).map_err(|_| HeaderError::CannotSerialize("invalid content type"))?;
+
+        data_writer.write_u8(version_with_flags);
+        // writing pad to returning bytes vec
+        data_writer.write_u8(0);
+        data_writer.write_u16_be(content_type_number);
+
+        Ok(data_writer.into_vec())
+    }
+}
+
+mod header_content {
+    use std::convert::TryFrom;
+    use std::num::TryFromIntError;
+
+    use num_enum::{FromPrimitive, IntoPrimitive};
+
+    /// The lowest 255 message types are reserved for cjdns/IPv6 packets.
+    /// AKA: packets where the IP address is within the FC00::/8 block.
+    /// Any packet sent in this way will have the IPv6 header deconstructed and this
+    /// field will come from the nextHeader field in the IPv6 header.
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, FromPrimitive, IntoPrimitive)]
+    #[repr(u32)]
+    pub enum ContentType {
+        Ip6Hop = 0,
+        Ip6Icmp = 1,
+        Ip6Igmp = 2,
+        Ip6Ipv4 = 4,
+        Ip6Tcp = 6,
+        Ip6Egp = 8,
+        Ip6Pup = 12,
+        Ip6Udp = 17,
+        Ip6Idp = 22,
+        Ip6Tp = 29,
+        Ip6Dccp = 33,
+        Ip6Ipv6 = 41,
+        Ip6Rsvp = 46,
+        Ip6Gre = 47,
+        Ip6Esp = 50,
+        Ip6Ah = 51,
+        Ip6Icmpv6 = 58,
+        Ip6Mtp = 92,
+        Ip6Beetph = 94,
+        Ip6Encap = 98,
+        Ip6Pim = 103,
+        Ip6Comp = 108,
+        Ip6Sctp = 132,
+        Ip6Udplite = 136,
+        Ip6Max = 255,
+
+        /// Bencoded inter-router DHT messages
+        Cjdht = 256,
+        Iptun = 257,
+
+        /// Reserved for future allocation
+        Reserved = 258,
+        ReservedMax = 0x7fff,
+
+        /// Content types in the AVAILABLE range are not defined and can be used
+        /// like port numbers for subsystems of cjdns to communicate with subsystems within
+        /// cjdns on other machines, providing they first agree on which numbers to use via
+        /// CTRL messages
+        Available = 0x8000,
+
+        /// This content type will never appear in the wild, it represents unencrypted control frames.
+        Ctrl = 0xffff + 1,
+
+        // TODO normally, as a default we should return Unknown(u16)
+        #[num_enum(default)]
+        Max = 0xffff + 2,
+    }
+
+    pub(super) fn to_u16(content_type: ContentType) -> Result<u16, TryFromIntError> {
```

Comment:
Better make this fn a method of `ContentType`

Reply:
And maybe rename to `try_to_u16`

Reply:
And no need to return that exact `TryFromIntError`, it is useless outside of this method. Just return `Result<u16, ()>` or `Option<u16>`.

Reply:
took it into consideration